tiledb/sm/global_state/tbb_state.cc

```diff
@@ -33,44 +33,62 @@
 
 #include "tiledb/sm/global_state/tbb_state.h"
 #include "tiledb/sm/config/config.h"
+#include "tiledb/sm/misc/thread_pool.h"
 
-#ifdef HAVE_TBB
+#include <cassert>
+#include <sstream>
 
+#ifdef HAVE_TBB
 #include <tbb/task_scheduler_init.h>
-#include <cassert>
 #include <cstdlib>
 #include <memory>
-#include <sstream>
+#endif
 
 namespace tiledb {
 namespace sm {
 namespace global_state {
 
-/** The TBB scheduler, used for controlling the number of TBB threads. */
-static std::unique_ptr<tbb::task_scheduler_init> tbb_scheduler_;
-
-/** The number of TBB threads the scheduler was configured with **/
-int tbb_nthreads_;
-
-Status init_tbb(const Config* config) {
-  int nthreads;
+Status get_nthreads(const Config* const config, int* const nthreads) {
   if (!config) {
-    nthreads = std::strtol(Config::SM_NUM_TBB_THREADS.c_str(), nullptr, 10);
+    *nthreads = std::strtol(Config::SM_NUM_TBB_THREADS.c_str(), nullptr, 10);
   } else {
     bool found = false;
-    RETURN_NOT_OK(config->get<int>("sm.num_tbb_threads", &nthreads, &found));
+    RETURN_NOT_OK(config->get<int>("sm.num_tbb_threads", nthreads, &found));
     assert(found);
   }
 
-  if (nthreads == tbb::task_scheduler_init::automatic) {
-    nthreads = tbb::task_scheduler_init::default_num_threads();
+#ifdef HAVE_TBB
+  if (*nthreads == tbb::task_scheduler_init::automatic) {
+    *nthreads = tbb::task_scheduler_init::default_num_threads();
+  }
+#else
+  if (*nthreads <= 0) {
+    *nthreads = std::thread::hardware_concurrency();
   }
-  if (nthreads < 1) {
+#endif
+
+  if (*nthreads < 1) {
     std::stringstream msg;
     msg << "TBB thread runtime must be initialized with >= 1 threads, got: "
```

Comment:
Remove mention of "TBB" here? We may have to make a global grep for TBB and correct in case multi-threading is too specific to TBB. That will allow us to easily fix all references again if we choose to sunset TBB for good.

Reply:
Leaving as-is based on previous discussion. To be changed in the future.